dartagnan/src/main/java/com/dat3m/dartagnan/wmm/processing/MergeEquivalentRelations.java

```diff
@@ -0,0 +1,139 @@
+package com.dat3m.dartagnan.wmm.processing;
+
+import com.dat3m.dartagnan.wmm.Constraint;
+import com.dat3m.dartagnan.wmm.Definition;
+import com.dat3m.dartagnan.wmm.Relation;
+import com.dat3m.dartagnan.wmm.Wmm;
+import com.dat3m.dartagnan.wmm.definition.CartesianProduct;
+import com.dat3m.dartagnan.wmm.definition.Fences;
+import com.dat3m.dartagnan.wmm.definition.SameScope;
+import com.dat3m.dartagnan.wmm.definition.SetIdentity;
+import com.dat3m.dartagnan.wmm.utils.ConstraintCopier;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import java.util.IdentityHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.function.Predicate;
+
+public class MergeEquivalentRelations implements WmmProcessor {
+
+    private static final Logger logger = LogManager.getLogger(MergeEquivalentRelations.class);
+
+    private MergeEquivalentRelations() {
+    }
+
+    public static MergeEquivalentRelations newInstance() {
+        return new MergeEquivalentRelations();
+    }
+
+    @Override
+    public void run(Wmm wmm) {
+        final Map<Relation, Relation> rel2class = computeEquivalenceClasses(wmm);
+        mergeEquivalentRelations(wmm, rel2class);
+    }
+
+    private Map<Relation, Relation> computeEquivalenceClasses(Wmm wmm) {
+        final Map<Relation, Relation> rel2class = new IdentityHashMap<>();
+        final Predicate<Relation> hasNontrivialClass = r -> rel2class.get(r) != r;
+        final List<Relation> relations = List.copyOf(wmm.getRelations());
+        relations.forEach(r -> rel2class.put(r, r));
+
+        // TODO: Could be done more efficiently by computing equivalence bottom-up along a stratification
+        //  and reusing previous equivalence queries.
+        //  However, even on large memory models like LKMM, this code takes less than 10ms.
+        for (int i = 0; i < relations.size(); i++) {
+            final Relation r1 = relations.get(i);
+            if (hasNontrivialClass.test(r1)) {
+                continue;
+            }
+            for (int j = i + 1; j < relations.size(); j++) {
+                final Relation r2 = relations.get(j);
+                if (!hasNontrivialClass.test(r2) && areEquivalent(r1, r2)) {
+                    rel2class.put(r2, r1);
+                }
+            }
+        }
+
+        return rel2class;
+    }
+
+    private boolean areEquivalent(Relation r1, Relation r2) {
+        if (r1 == r2)  {
+            return true;
+        }
+        final Definition def1 = r1.getDefinition();
+        final Definition def2 = r2.getDefinition();
+        if (def1.getClass() != def2.getClass()) {
+            return false;
+        }
+
+        if (r1.isRecursive() || r2.isRecursive()) {
+            // For now, we return false for different recursive relations.
```

Comment:
If you have some idea of how this could be changed in the future, please add a TODO ... having the keywork might help when someone has some free time and looks for low-hanging fruits to improve the code

Reply:
Then I would remove the "for now" (and "TODO") all together